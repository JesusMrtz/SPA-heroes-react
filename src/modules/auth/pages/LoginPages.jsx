import { useContext } from "react";
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context";

export const LoginPages = () => {
  const navigate = useNavigate();
  /** Damos acceso al context */
  const { login } = useContext(AuthContext);

  function onLogin() {
    login( 'Jesus Mart√≠nez' );

    navigate('/', {
      replace: true
    });
  }

  return (
    <>
      <h1>Login Page</h1>
      <hr />
      <button className="btn btn-primary" onClick={ onLogin }>Login</button>
    </>
  )
}
