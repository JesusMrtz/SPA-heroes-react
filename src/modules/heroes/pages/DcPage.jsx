import { HeroListComponent } from "../components"

export const DcPage = () => {
  return (
    <>
      <h1>DC Page</h1>
      <hr />
      <HeroListComponent publisher="DC Comics" />
    </>
  )
}
