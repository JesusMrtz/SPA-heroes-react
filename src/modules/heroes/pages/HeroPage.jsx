import { useEffect, useMemo } from "react";
import { useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers";


export const HeroPage = () => {
  const navigate = useNavigate();
  /** Obtener los parÃ¡metros de la URL */
  const { id } = useParams();
  const hero =  useMemo(() => getHeroById(id), [ id ]);

  
  useEffect(() => {
    if ( !hero ) {
      return onNavigateBack();
    }
  }, []);

  if ( !hero ) {
    return onNavigateBack();
  }
  
  const heroImageURL = `/assets/img/heroes/${ id }.jpg`;

  function onNavigateBack() {
    navigate(-1);
    return;
  }

  return (
    <>
      <div className="row mt-5">
        <div className="col-12 col-sm-4 mb-2 mb-sm-0">
          <img src={ heroImageURL } alt={ hero.superhero } className="img-thumbnail animate__animated animate__fadeInLeft" loading="lazy" />
        </div>
        <div className="col text-center animate__animated animate__fadeIn">
          <h1>{ hero.superhero }</h1>
          <ul className="list-group list-group-flush">
            <li className="list-group-item">
              <strong>Alter ego: </strong> { hero.alter_ego }
            </li>
            <li className="list-group-item">
              <strong>Publisher: </strong> { hero.publisher }
            </li>
            <li className="list-group-item">
              <strong>First appearance: </strong> { hero.first_appearance }
            </li>
          </ul>
          <h2 className="mt-3">Characters</h2>
          <p>
            { hero.characters }
          </p>
          <button className="btn btn-outline-primary" onClick={ onNavigateBack } >
            Regresar
          </button>
        </div>
      </div>
    </>
  )
}