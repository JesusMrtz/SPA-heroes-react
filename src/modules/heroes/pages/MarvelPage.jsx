import { HeroListComponent } from "../components";

export const MarvelPage = () => {
  return (
    <>
      <h1>Marvel Page</h1>
      <hr />
      <HeroListComponent publisher="Marvel Comics" />
    </>
  );
}
