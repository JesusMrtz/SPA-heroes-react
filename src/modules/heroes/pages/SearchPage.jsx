import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../shared/hooks/useForm"
import queryString from 'query-string';
import { getHeroesByName } from "../helpers";
import { HeroItemComponent } from "../components/HeroItemComponent";


export const SearchPage = () => {
  const navigate = useNavigate();
  /** se usa el useLocation para leer los query parameters ?q */
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);
  const heroes = getHeroesByName(q);

  const { searchText, onInputChange } = useForm({
    searchText: q
  });

  const showErrorAlert = (q.length && !heroes.length)

  function onSearchSubmit(e) {
    e.preventDefault();
    const value = searchText.trim();
    navigate(`?q=${ value }`);
  }

  return (
    <>
      <h1>Buscador de heróes</h1>
      <hr />
      <div className="row">
        <div className="col-12 col-sm-5">
          <h2>Buscando</h2>
          <hr />
          <form onSubmit={ onSearchSubmit } data-testid='form'>
            <input type="text" placeholder="Buscar heroe" className="form-control" name="searchText" autoComplete="off" value={ searchText } onChange={ onInputChange } />
            <button type="submit" className="btn btn-outline-primary mt-1">Buscar</button>  
          </form>
        </div>
        <div className="col-12 col-sm-7">
          <h2>Results</h2>
          <hr />
          <div className="alert alert-primary animate__animated animate__fadeIn" style={{ display: q !== '' ? 'none' : 'block' }}>
            Buscar un heroe
          </div>
          <div className="alert alert-danger animate__animated animate__fadeIn" style={{ display: showErrorAlert ? 'block': 'none' }} data-testid='alert-danger'>
            No se encontró el heroe <strong>{ q }</strong>
          </div>
          {
            heroes.map(hero => (
              <div className="col-12 mb-4" key={ hero.id }>
                <HeroItemComponent hero={ hero } />
              </div>
            ))
          }
        </div>
      </div>
    </>
  )
}
