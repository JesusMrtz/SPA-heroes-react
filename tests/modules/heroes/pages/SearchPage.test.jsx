import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchPage } from "../../../../src/modules/heroes/pages/SearchPage";


const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
}));


describe(`Pruebas en page <SearchPage />`, () => {
    it(`Debe de mostrarse correctamente con los valores por defecto`, () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();
    });

    beforeEach(() => jest.clearAllMocks() );


    it(`Debe de mostrarse a Batman y el input con el query-string`, () => {
        render(
            <MemoryRouter initialEntries={ ['/superheroes/search?q=batman'] }>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman');

        const imgBatman = screen.getByRole('img');
        expect( imgBatman.src ).toContain('batman');
    });


    it(`Debe de mostrar un error si no se encuentra el heroe`, () => {
        render(
            <MemoryRouter initialEntries={ ['/superheroes/search?q=karely'] }>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('karely');

        const errorDiv = screen.getByTestId('alert-danger');
        expect( errorDiv.style.display ).toBe('block');
    });


    it(`Debe de llamar al navigate a la pantalla nueva`, async () => {
        render(
            <MemoryRouter initialEntries={ ['/superheroes/search'] }>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        await fireEvent.change(input, { target: { name: 'searchText', value: 'batman' } });

        const formInput = screen.getByTestId('form');
        await fireEvent.submit( formInput );

        expect( mockUseNavigate ).toHaveBeenCalled();
        expect( mockUseNavigate ).toHaveBeenCalledWith('?q=batman');
    });
});