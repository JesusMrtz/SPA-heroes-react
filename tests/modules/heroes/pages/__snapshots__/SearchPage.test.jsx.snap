// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en page <SearchPage /> Debe de mostrarse correctamente con los valores por defecto 1`] = `
<div>
  <h1>
    Buscador de heróes
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-12 col-sm-5"
    >
      <h2>
        Buscando
      </h2>
      <hr />
      <form
        data-testid="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Buscar heroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-12 col-sm-7"
    >
      <h2>
        Results
      </h2>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
        style="display: block;"
      >
        Buscar un heroe
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        data-testid="alert-danger"
        style="display: none;"
      >
        No se encontró el heroe 
        <strong />
      </div>
    </div>
  </div>
</div>
`;
